File: android/app/src/main/java/org/nmp/android/TypedHomeFeedDecoder.kt
package org.nmp.android

import android.util.Log
import nmp.content.ContentTreeWire as FbContentTreeWire
import nmp.content.WireNodeKind
import nmp.content.WireNostrUriKind
import nmp.content.PlaceholderReason as FbPlaceholderReason
import nmp.content.RenderMode as FbRenderMode
import nmp.feed.FeedWindow
import nmp.nip01.OpFeedSnapshot
import nmp.nip01.ReplyAttribution
import nmp.nip01.RootCard
import nmp.nip01.TimelineEventCard
import org.nmp.android.model.ChirpEventCard
import org.nmp.android.model.ChirpOpFeedSnapshot
import org.nmp.android.model.ChirpReplyAttribution
import org.nmp.android.model.ChirpRootCard
import org.nmp.android.model.ContentTreeWire
import org.nmp.android.model.ContentWireNode
import org.nmp.android.model.TimelineWindowCursor
import org.nmp.android.model.TimelineWindowPage
import org.nmp.android.model.WireNostrUri
import java.nio.ByteBuffer
import java.nio.ByteOrder

private const val TAG = "TypedHomeFeedDecoder"

/** Sentinel: `event_kind = u32::MAX` means `None` (mirrors `EVENT_KIND_NONE` in typed_fb.rs). */
private const val EVENT_KIND_NONE: UInt = UInt.MAX_VALUE

/** Sentinel: `ordered_start = -1` means unordered list (mirrors `ORDERED_START_NONE`). */
private const val ORDERED_START_NONE: Long = -1L

/**
 * Decodes the typed `nmp.feed.home` sidecar from a FlatBuffers `NOFS` buffer
 * (ADR-0038 Stage T4 / B4 — V-85 complete) into a [ChirpOpFeedSnapshot].
 *
 * ADR-0037 introduced typed FlatBuffers runtime projections carried alongside
 * the generic snapshot `payload`. The authorized pilot is `nmp.feed.home`,
 * whose OP-centric view is the nmp-feed `RootFeedSnapshot<TimelineEventCard,
 * Nip10ReplyAttribution>` (`schema_id = "nmp.nip01.opfeed"`, `file_identifier
 * = "NOFS"`). The retired NFTS descriptor (`nmp.nip01.timeline`) is no longer
 * preferred — an `NFTS`-tagged entry is treated as unrecognized and falls
 * through to the generic projection (ADR-0037 Commitment 4).
 *
 * Every entry point falls back gracefully — it returns `null` when the
 * projection is absent, carries the wrong schema id, or cannot be verified as
 * a well-formed `NOFS` buffer. Hosts treat `null` as "no typed feed available"
 * and keep rendering the generic snapshot (ADR-0037 Commitment 4 permanent
 * fallback — the generic `Value` path is never removed).
 *
 * V-85 adds the native Kotlin NFCT decoder (`decodeContentTree`) so
 * [ChirpEventCard.contentTree] is now populated from the embedded
 * `content_tree_bytes` sub-buffer inside each [TimelineEventCard]. The typed
 * path is now the live preferred path; `KernelModel.decodeUpdate` wires it.
 */
object TypedHomeFeedDecoder {

    /** Projection key published by the kernel (`TypedProjection.key`). */
    const val PROJECTION_KEY = "nmp.feed.home"

    /** Schema id carried in `TypedPayload.schema_id` for the NOFS wire. */
    const val SCHEMA_ID = "nmp.nip01.opfeed"

    /** FlatBuffers `file_identifier` for `OpFeedSnapshot`. */
    const val FILE_IDENTIFIER = "NOFS"

    /**
     * Extract and decode the `nmp.feed.home` typed payload from a list of
     * [TypedProjectionEnvelope]s lifted off a snapshot frame.
     *
     * Mirrors iOS `TypedHomeFeedDecoder.decode(from:)`. Returns `null` (no
     * typed feed) when the matching NOFS entry is absent or empty.
     */
    fun decode(projections: List<TypedProjectionEnvelope>): ChirpOpFeedSnapshot? {
        val projection = projections.firstOrNull {
            it.key == PROJECTION_KEY && it.schemaId == SCHEMA_ID
        } ?: return null
        if (projection.payload.isEmpty()) return null
        return decode(projection.payload)
    }

    /**
     * Decode a raw `NOFS` FlatBuffers buffer into a [ChirpOpFeedSnapshot].
     *
     * Mirrors iOS `TypedHomeFeedDecoder.decode(bytes:)`. Verifies the
     * file_identifier before reading any fields; returns `null` on any parse
     * error so the host falls back to the generic projection.
     */
    fun decode(bytes: ByteArray): ChirpOpFeedSnapshot? {
        if (bytes.isEmpty()) return null
        return try {
            val bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
            if (!OpFeedSnapshot.OpFeedSnapshotBufferHasIdentifier(bb)) {
                Log.e(TAG, "NOFS file_identifier missing (${bytes.size} bytes)")
                return null
            }
            val snapshot = OpFeedSnapshot.getRootAsOpFeedSnapshot(bb)
            val cards = buildList {
                for (i in 0 until snapshot.cardsLength) {
                    val root = snapshot.cards(i) ?: continue
                    add(makeRootCard(root))
                }
            }
            val page = if (snapshot.hasPage) decodePage(snapshot) else null
            ChirpOpFeedSnapshot(cards = cards, page = page)
        } catch (e: Exception) {
            Log.e(TAG, "NOFS decode error: ${e.message} bytes=${bytes.size}")
            null
        }
    }

    // ── Card mapping ─────────────────────────────────────────────────────────

    private fun makeRootCard(root: RootCard): ChirpRootCard {
        val attribution = buildList {
            for (i in 0 until root.attributionLength) {
                val entry = root.attribution(i) ?: continue
                add(makeAttribution(entry))
            }
        }
        return ChirpRootCard(card = makeCard(root.card), attribution = attribution)
    }

    private fun makeCard(card: TimelineEventCard?): ChirpEventCard {
        // Decode the embedded NFCT content-tree sub-buffer when present.
        // `contentTreeBytesAsByteBuffer` returns null when the vector field
        // is absent (length == 0); guard matches the NFWM pattern in decodePage.
        val contentTree: ContentTreeWire? = if ((card?.contentTreeBytesLength ?: 0) > 0) {
            card?.contentTreeBytesAsByteBuffer?.let { buf ->
                buf.order(ByteOrder.LITTLE_ENDIAN)
                decodeContentTree(buf)
            }
        } else null
        return ChirpEventCard(
            id = card?.id ?: "",
            authorPubkey = card?.authorPubkey ?: "",
            kind = (card?.kind ?: 0u).toInt(),
            createdAt = (card?.createdAt ?: 0UL).toLong(),
            content = card?.content ?: "",
            contentTree = contentTree,
            // ADR-0032: `has_*` companion bool distinguishes "absent (no kind:0
            // yet)" from "present empty string".
            authorDisplayName = if (card?.hasAuthorDisplayName == true) card.authorDisplayName else null,
            authorPictureUrl = if (card?.hasAuthorPictureUrl == true) card.authorPictureUrl else null,
            contentPreview = card?.contentPreview ?: "",
        )
    }

    private fun makeAttribution(entry: ReplyAttribution): ChirpReplyAttribution {
        return ChirpReplyAttribution(
            authorPubkey = entry.authorPubkey ?: "",
            authorDisplayName = if (entry.hasAuthorDisplayName) entry.authorDisplayName else null,
            authorPictureUrl = if (entry.hasAuthorPictureUrl) entry.authorPictureUrl else null,
            replyEventId = entry.replyEventId ?: "",
            replyCreatedAt = entry.replyCreatedAt,
        )
    }

    // ── NFCT content-tree sub-buffer decoder ──────────────────────────────────

    /**
     * Decode an embedded `NFCT` FlatBuffers sub-buffer into a [ContentTreeWire].
     *
     * Verifies the `"NFCT"` file identifier before reading any fields — the
     * same guard used for the `"NFWM"` feed-window buffer in [decodePage].
     * Returns `null` on absent identifier, empty buffer, or any parse error
     * (D1: fail closed, never crash). Raw values only — no display helpers,
     * no relative time, no short-hex formatting (D8).
     *
     * NFCT is generated by `nmp-content::wire::typed_fb::encode_content_tree`
     * (schema: `crates/nmp-content/schema/content_tree.fbs`).
     * All 22 [WireNodeKind] variants are handled:
     * Text(0) Mention(1) EventRef(2) Hashtag(3) Url(4) Media(5) Emoji(6)
     * Invoice(7→Placeholder) Heading(8) Paragraph(9) BlockQuote(10)
     * CodeBlock(11) List(12) Rule(13) Emphasis(14) Strong(15) InlineCode(16)
     * Link(17) Image(18) SoftBreak(19) HardBreak(20) Placeholder(21).
     * Invoice is mapped to [ContentWireNode.PlaceholderNode] to match the
     * generic JSON fallback path (the Kotlin model has no InvoiceNode).
     */
    private fun decodeContentTree(buf: ByteBuffer): ContentTreeWire? {
        if (!FbContentTreeWire.ContentTreeWireBufferHasIdentifier(buf)) {
            Log.e(TAG, "NFCT file_identifier missing")
            return null
        }
        return try {
            val tree = FbContentTreeWire.getRootAsContentTreeWire(buf)
            val nodes = buildList {
                for (i in 0 until tree.nodesLength) {
                    val node = tree.nodes(i) ?: continue
                    add(decodeWireNode(node))
                }
            }
            val roots = buildList {
                for (i in 0 until tree.rootsLength) {
                    add(tree.roots(i).toInt())
                }
            }
            val mode = renderModeFromFb(tree.mode)
            ContentTreeWire(nodes = nodes, roots = roots, mode = mode)
        } catch (e: Exception) {
            Log.e(TAG, "NFCT decode error: ${e.message}")
            null
        }
    }

    /**
     * Map a single FlatBuffers [nmp.content.WireNode] to [ContentWireNode].
     *
     * Dispatch is on `kind` only — several variants share field names (`text`,
     * `children`); the discriminator is the sole authority on which fields are
     * meaningful, mirroring the Rust decode in `typed_fb.rs::decode_node`.
     */
    private fun decodeWireNode(node: nmp.content.WireNode): ContentWireNode {
        return when (node.kind) {
            WireNodeKind.Text -> ContentWireNode.TextNode(node.text.orEmpty())
            WireNodeKind.Mention -> ContentWireNode.MentionNode(decodeNostrUri(node))
            WireNodeKind.EventRef -> ContentWireNode.EventRefNode(decodeNostrUri(node))
            WireNodeKind.Hashtag -> ContentWireNode.HashtagNode(node.tag.orEmpty())
            WireNodeKind.Url -> ContentWireNode.UrlNode(node.url.orEmpty())
            WireNodeKind.Media -> ContentWireNode.MediaNode(
                urls = buildList { for (i in 0 until node.mediaUrlsLength) add(node.mediaUrls(i).orEmpty()) },
                mediaKind = mediaKindString(node.mediaKind),
            )
            WireNodeKind.Emoji -> ContentWireNode.EmojiNode(
                shortcode = node.shortcode.orEmpty(),
                url = node.emojiUrl,
            )
            // Invoice: no InvoiceNode in the Kotlin model; map to Placeholder to
            // match the generic JSON path (`ContentWireNodeSerializer` else branch).
            WireNodeKind.Invoice -> ContentWireNode.PlaceholderNode
            WireNodeKind.Heading -> ContentWireNode.HeadingNode(
                level = node.level.toInt(),
                children = childrenList(node),
            )
            WireNodeKind.Paragraph -> ContentWireNode.ParagraphNode(childrenList(node))
            WireNodeKind.BlockQuote -> ContentWireNode.BlockQuoteNode(childrenList(node))
            WireNodeKind.CodeBlock -> ContentWireNode.CodeBlockNode(
                info = node.codeInfo,
                body = node.text.orEmpty(),
            )
            WireNodeKind.List -> {
                // ordered_start default in schema is -1 (ORDERED_START_NONE = unordered).
                val orderedStart: Long? = if (node.orderedStart == ORDERED_START_NONE) null else node.orderedStart
                ContentWireNode.ListNode(
                    orderedStart = orderedStart,
                    items = buildList {
                        for (i in 0 until node.listItemsLength) {
                            val item = node.listItems(i) ?: continue
                            add(buildList { for (j in 0 until item.childrenLength) add(item.children(j).toInt()) })
                        }
                    },
                )
            }
            WireNodeKind.Rule -> ContentWireNode.RuleNode
            WireNodeKind.Emphasis -> ContentWireNode.EmphasisNode(childrenList(node))
            WireNodeKind.Strong -> ContentWireNode.StrongNode(childrenList(node))
            WireNodeKind.InlineCode -> ContentWireNode.InlineCodeNode(node.text.orEmpty())
            WireNodeKind.Link -> ContentWireNode.LinkNode(
                children = childrenList(node),
                href = node.href,
            )
            WireNodeKind.Image -> ContentWireNode.ImageNode(
                alt = node.alt.orEmpty(),
                // `url` field in the schema encodes `src` for Image nodes
                // (see encode_node in typed_fb.rs: `args.url = src`).
                // `imgTitle` is dropped — the Kotlin model has no title field.
                src = node.url,
            )
            WireNodeKind.SoftBreak -> ContentWireNode.SoftBreakNode
            WireNodeKind.HardBreak -> ContentWireNode.HardBreakNode
            WireNodeKind.Placeholder -> ContentWireNode.PlaceholderNode
            else -> ContentWireNode.PlaceholderNode // forward-compat: unknown kind
        }
    }

    /** Decode the `nostr_uri` sub-table of a Mention or EventRef node. */
    private fun decodeNostrUri(node: nmp.content.WireNode): WireNostrUri {
        val fb = node.nostrUri ?: return WireNostrUri()
        val kind = when (fb.kind) {
            WireNostrUriKind.Profile -> "profile"
            WireNostrUriKind.Event -> "event"
            WireNostrUriKind.Address -> "address"
            else -> "profile"
        }
        // event_kind uses EVENT_KIND_NONE (u32::MAX) as the None sentinel.
        val eventKind: Int? = if (fb.eventKind == EVENT_KIND_NONE) null else fb.eventKind.toInt()
        return WireNostrUri(
            uri = fb.uri.orEmpty(),
            kind = kind,
            primaryId = fb.primaryId.orEmpty(),
            relays = buildList { for (i in 0 until fb.relaysLength) add(fb.relays(i).orEmpty()) },
            author = fb.author,
            eventKind = eventKind,
        )
    }

    /** Collect the `children [uint32]` vector into an `Int` list. */
    private fun childrenList(node: nmp.content.WireNode): List<Int> =
        buildList { for (i in 0 until node.childrenLength) add(node.children(i).toInt()) }

    /**
     * Map a FlatBuffers `media_kind` uint8 to the string the Kotlin model
     * carries (matches `MediaKind` serde PascalCase — no `rename_all` in Rust).
     * Image=0, Video=1, Audio=2.
     */
    private fun mediaKindString(v: UByte): String = when (v) {
        0u.toUByte() -> "Image"
        1u.toUByte() -> "Video"
        2u.toUByte() -> "Audio"
        else -> "Image" // forward-compat default
    }

    /**
     * Map a FlatBuffers `RenderMode` byte to the string the Kotlin model uses.
     * Schema: Auto=0, Markdown=1, Text=2. Text maps to "Plain" — the Rust wire
     * encoder maps `RenderMode::Plain → fb::RenderMode::Text` (value 2).
     */
    private fun renderModeFromFb(v: UByte): String = when (v) {
        nmp.content.RenderMode.Auto -> "Auto"
        nmp.content.RenderMode.Markdown -> "Markdown"
        nmp.content.RenderMode.Text -> "Plain"
        else -> "Auto"
    }

    // ── Feed-window (NFWM) sub-buffer → page ──────────────────────────────────

    /**
     * Decode the embedded `feed_window_bytes` (`NFWM`) sub-buffer and map its
     * `FeedPage` to the [TimelineWindowPage] the renderer paginates on. Returns
     * `null` when the window is absent, malformed, or carries no page (the
     * generic decoder likewise ignores `metrics`, so this maps page only).
     */
    private fun decodePage(snapshot: OpFeedSnapshot): TimelineWindowPage? {
        if (snapshot.feedWindowBytesLength == 0) return null
        // `feedWindowBytesAsByteBuffer` is non-null for this generated table
        // (the `[ubyte]` accessor); the length guard above rules out an empty
        // window, so the slice is a well-formed embedded NFWM buffer.
        val windowBuffer = snapshot.feedWindowBytesAsByteBuffer
        windowBuffer.order(ByteOrder.LITTLE_ENDIAN)
        if (!FeedWindow.FeedWindowBufferHasIdentifier(windowBuffer)) return null
        val window = FeedWindow.getRootAsFeedWindow(windowBuffer)
        val page = window.page ?: return null
        val cursor = page.nextCursor?.let { raw ->
            val id = raw.id ?: return@let null
            TimelineWindowCursor(createdAt = raw.createdAt, id = id)
        }
        return TimelineWindowPage(
            limit = page.limit,
            nextCursor = cursor,
            hasMore = page.hasMore,
            totalBlocks = page.totalBlocks,
        )
    }
}

File: apps/chirp/ios/Chirp/Bridge/TypedHomeFeedDecoder.swift
import Foundation
import FlatBuffers

/// Decodes the typed `nmp.feed.home` sidecar from a FlatBuffers `NOFS` buffer
/// (ADR-0038 Stage T4) into the SAME `OpFeedSnapshot` model the generic
/// `Value` projection path produces, so `HomeFeedView` renders either source
/// identically.
///
/// ADR-0037 introduced typed FlatBuffers runtime projections carried alongside
/// the generic snapshot `payload`. The authorized pilot is `nmp.feed.home`,
/// whose OP-centric view is the nmp-feed `RootFeedSnapshot<TimelineEventCard,
/// Nip10ReplyAttribution>` (`schema_id = "nmp.nip01.opfeed"`, `file_identifier
/// = "NOFS"`). The retired NFTS descriptor (`nmp.nip01.timeline`) is no longer
/// preferred — an `NFTS`-tagged entry is treated as unrecognized and falls
/// through to the generic projection (ADR-0037 Commitment 4).
///
/// Every entry point falls back gracefully — it returns `nil` when the
/// projection is absent, carries the wrong schema id, or cannot be verified as
/// a well-formed `NOFS` buffer. Hosts treat `nil` as "no typed feed available"
/// and keep rendering the generic snapshot.
///
/// CONSUMER STATUS — the typed path is now LIVE. `TypedHomeFeedDecoder` is
/// wired into `KernelBridge.decodeFlatBuffer` as the PREFERRED decode path;
/// `KernelModel.modularTimeline` returns the typed result when available,
/// falling back to the generic `Value` path when the typed decode returns nil
/// (ADR-0037 Commitment 4). The `NFCT` content-tree sub-buffer and the
/// `NoteRelationCounts` typed sub-table are both fully decoded in Swift using
/// the generated `nmp_content_*` FlatBuffers accessors.
enum TypedHomeFeedDecoder {
    /// Projection key as published by the kernel (`TypedProjection.key`).
    static let projectionKey = "nmp.feed.home"
    /// Schema id carried in `TypedPayload.schema_id` for the NOFS wire.
    static let schemaId = "nmp.nip01.opfeed"
    /// FlatBuffers `file_identifier` for `OpFeedSnapshot`.
    static let fileIdentifier = "NOFS"
    /// FlatBuffers `file_identifier` for the embedded `FeedWindow` sub-buffer.
    static let feedWindowFileIdentifier = "NFWM"
    /// FlatBuffers `file_identifier` for the embedded `ContentTreeWire` sub-buffer.
    static let contentTreeFileIdentifier = "NFCT"

    /// Extract and decode the `nmp.feed.home` typed payload from a set of typed
    /// projection envelopes lifted off a snapshot frame.
    static func decode(from projections: [TypedProjectionEnvelope]) -> OpFeedSnapshot? {
        guard let projection = projections.first(where: {
            $0.key == projectionKey && $0.schemaId == schemaId
        }), !projection.payload.isEmpty else {
            return nil
        }
        return decode(bytes: projection.payload)
    }

    /// Decode a raw `NOFS` FlatBuffers buffer into the Swift feed model.
    ///
    /// Uses `getRoot` (unchecked) because this buffer arrives over a trusted
    /// in-process FFI boundary from the Rust kernel — running the O(N)
    /// FlatBuffers Verifier on every 4 Hz frame is pure wasted CPU. The
    /// `!bytes.isEmpty` guard above is the only presence check needed; a
    /// gross wiring error would surface as empty/default field values from
    /// the glue layer rather than a crash.
    static func decode(bytes: Data) -> OpFeedSnapshot? {
        guard !bytes.isEmpty else { return nil }
        var buffer = ByteBuffer(data: bytes)
        let snapshot: nmp_nip01_OpFeedSnapshot = getRoot(byteBuffer: &buffer)

        let cards = snapshot.cards.map(makeRootCard)
        let page = snapshot.hasPage ? decodePage(snapshot) : nil
        return OpFeedSnapshot(cards: cards, page: page)
    }

    // ── Card mapping ─────────────────────────────────────────────────────────

    private static func makeRootCard(_ root: nmp_nip01_RootCard) -> ChirpRootCard {
        ChirpRootCard(
            card: makeCard(root.card),
            attribution: root.attribution.map(makeAttribution)
        )
    }

    private static func makeCard(_ card: nmp_nip01_TimelineEventCard?) -> ChirpEventCard {
        // Decode the embedded NFCT content-tree sub-buffer. Returns nil when
        // the buffer is absent or malformed — ChirpEventCard.contentTree is
        // Optional, so nil is valid and the renderer falls back to plain content.
        let contentTree: ContentTreeWire? = card.flatMap { decodeContentTree($0) }

        // Map the typed NoteRelationCounts sub-table. Absent == nil.
        let relationCounts: NoteRelationCounts? = card?.relationCounts.flatMap(makeRelationCounts)

        return ChirpEventCard(
            id: card?.id ?? "",
            authorPubkey: card?.authorPubkey ?? "",
            kind: card?.kind ?? 0,
            createdAt: card?.createdAt ?? 0,
            content: card?.content ?? "",
            contentTree: contentTree,
            relationCounts: relationCounts,
            // ADR-0032: `has_*` companion bool distinguishes "absent (no kind:0
            // yet)" from "present empty string".
            authorDisplayName: optionalString(card?.authorDisplayName, present: card?.hasAuthorDisplayName ?? false),
            authorPictureUrl: optionalString(card?.authorPictureUrl, present: card?.hasAuthorPictureUrl ?? false),
            contentPreview: card?.contentPreview ?? "",
            relayProvenance: card?.relayProvenance.compactMap { $0 } ?? [],
            // Typed NIP-18 repost signal (D0): the kernel stamps `reposted_by`
            // when a kind:6 repost superseded the original note. We must read
            // this typed table rather than re-derive `kind == 6` in the view —
            // a repost card carries the *original* note's kind, not 6.
            isRepost: card?.repostedBy != nil
        )
    }

    private static func makeAttribution(_ entry: nmp_nip01_ReplyAttribution) -> ChirpReplyAttribution {
        // ADR-0032 / #1493: flat mirrors removed; read nested authorDisplay.
        let display = entry.authorDisplay
        return ChirpReplyAttribution(
            authorPubkey: entry.authorPubkey ?? "",
            authorDisplayName: optionalString(display?.name, present: display?.hasName ?? false),
            authorPictureUrl: optionalString(display?.pictureUrl, present: display?.hasPictureUrl ?? false),
            replyEventId: entry.replyEventId ?? "",
            replyCreatedAt: entry.replyCreatedAt
        )
    }

    // ── Feed-window (NFWM) sub-buffer → page ──────────────────────────────────

    /// Decode the embedded `feed_window_bytes` (`NFWM`) sub-buffer and map its
    /// `FeedPage` to the Swift `TimelineWindowPage` the renderer paginates on.
    /// Returns `nil` when the window is absent, malformed, or carries no page
    /// (the generic decoder likewise ignores `metrics`, so this maps page only).
    private static func decodePage(_ snapshot: nmp_nip01_OpFeedSnapshot) -> TimelineWindowPage? {
        // `withUnsafePointerToFeedWindowBytes` returns `T?` where the closure
        // result is itself `Data?`, so the call yields `Data??`. A `nil` window
        // (absent vector) or an empty slice both collapse to "no page".
        let windowData: Data? = snapshot.withUnsafePointerToFeedWindowBytes { pointer, count -> Data? in
            guard count > 0, let base = pointer.baseAddress else { return nil }
            return Data(bytes: base, count: count)
        } ?? nil
        guard let data = windowData, !data.isEmpty else { return nil }
        var windowBuffer = ByteBuffer(data: data)
        // Trusted in-process sub-buffer (embedded within a kernel-produced NOFS
        // frame); skip the O(N) Verifier walk.
        let window: nmp_feed_FeedWindow = getRoot(byteBuffer: &windowBuffer)
        guard let page = window.page else {
            return nil
        }
        // Explicit closure rather than `Optional.flatMap` to avoid any
        // Sequence/Optional overload-resolution ambiguity in the Swift FB types.
        let cursor: TimelineWindowCursor? = {
            guard let raw = page.nextCursor, let id = raw.id else { return nil }
            return TimelineWindowCursor(createdAt: raw.createdAt, id: id)
        }()
        return TimelineWindowPage(
            limit: UInt(clamping: page.limit),
            nextCursor: cursor,
            hasMore: page.hasMore,
            totalBlocks: UInt(clamping: page.totalBlocks)
        )
    }

    // ── NFCT content-tree sub-buffer → ContentTreeWire ────────────────────────

    /// Decode a raw `NFCT` FlatBuffers buffer directly into `ContentTreeWire`.
    ///
    /// Exposed as internal (not private) so tests can call it directly without
    /// needing a wrapping NOFS snapshot. For production use, prefer
    /// `decodeContentTree(_: nmp_nip01_TimelineEventCard)`.
    static func decodeContentTree(fromBytes bytes: Data) -> ContentTreeWire? {
        guard !bytes.isEmpty else { return nil }
        var treeBuffer = ByteBuffer(data: bytes)
        // Trusted in-process sub-buffer; skip the O(N) Verifier walk.
        let root: nmp_content_ContentTreeWire = getRoot(byteBuffer: &treeBuffer)
        let nodes = root.nodes.compactMap(decodeWireNode)
        let roots: [UInt32] = root.roots.map { $0 }
        let mode = renderModeString(root.mode)
        return ContentTreeWire(nodes: nodes, roots: roots, mode: mode)
    }

    /// Decode the embedded `content_tree_bytes` (`NFCT`) sub-buffer carried on
    /// a `TimelineEventCard` into the Swift `ContentTreeWire` model.
    ///
    /// Mirrors the `decodePage` pattern: extract the raw bytes via
    /// `withUnsafePointerToContentTreeBytes`, wrap in a `ByteBuffer`, then call
    /// `getRoot` (unchecked) to obtain the reader struct. The buffer is
    /// trusted (embedded within a kernel-produced NOFS frame); the O(N)
    /// FlatBuffers Verifier walk is skipped. Returns `nil` when the sub-buffer
    /// is absent — `contentTree` is Optional on `ChirpEventCard` so nil is
    /// valid and the renderer falls back to the plain `content` string.
    ///
    /// Raw values only — no display helpers (D11, display-separation audit
    /// 2026-05-25).
    static func decodeContentTree(_ card: nmp_nip01_TimelineEventCard) -> ContentTreeWire? {
        let treeData: Data? = card.withUnsafePointerToContentTreeBytes { pointer, count -> Data? in
            guard count > 0, let base = pointer.baseAddress else { return nil }
            return Data(bytes: base, count: count)
        } ?? nil
        guard let data = treeData, !data.isEmpty else { return nil }
        var treeBuffer = ByteBuffer(data: data)
        // Trusted in-process sub-buffer; skip the O(N) Verifier walk.
        let root: nmp_content_ContentTreeWire = getRoot(byteBuffer: &treeBuffer)
        let nodes = root.nodes.compactMap(decodeWireNode)
        let roots: [UInt32] = root.roots.map { $0 }
        let mode = renderModeString(root.mode)
        return ContentTreeWire(nodes: nodes, roots: roots, mode: mode)
    }

    /// Map the FlatBuffers `RenderMode` enum to the string the `ContentTreeWire`
    /// model carries. The JSON serde repr uses PascalCase (no `rename_all`):
    ///   Auto=0 → "Auto", Markdown=1 → "Markdown", Text=2 → "Plain"
    /// (the FB enum value `text` corresponds to Rust `RenderMode::Plain`).
    private static func renderModeString(_ mode: nmp_content_RenderMode) -> String? {
        switch mode {
        case .auto:     return "Auto"
        case .markdown: return "Markdown"
        case .text:     return "Plain"
        default:        return nil
        }
    }

    /// Decode one `WireNode` table into the Swift `NostrWireNode` enum.
    /// On any missing required field or unknown kind, returns nil (the node is
    /// silently dropped rather than aborting the whole tree decode — D1 partial).
    private static func decodeWireNode(_ node: nmp_content_WireNode) -> NostrWireNode? {
        switch node.kind {
        case .text:
            guard let text = node.text else { return nil }
            return .text(text)

        case .mention:
            guard let uri = decodeNostrUri(node.nostrUri) else { return nil }
            return .mention(uri)

        case .eventref:
            guard let uri = decodeNostrUri(node.nostrUri) else { return nil }
            return .eventRef(uri)

        case .hashtag:
            guard let tag = node.tag else { return nil }
            return .hashtag(tag)

        case .url:
            guard let url = node.url else { return nil }
            return .url(url)

        case .media:
            let urls = node.mediaUrls.compactMap { $0 }
            let kind = decodeMediaKind(node.mediaKind)
            return .media(urls: urls, kind: kind)

        case .emoji:
            guard let shortcode = node.shortcode else { return nil }
            return .emoji(shortcode: shortcode, url: node.emojiUrl)

        case .invoice:
            guard let payload = node.invoicePayload else { return nil }
            let invoice = decodeInvoice(kind: node.invoiceKind, payload: payload)
            return .invoice(invoice)

        case .heading:
            let children: [UInt32] = node.children.map { $0 }
            return .heading(level: node.level, children: children)

        case .paragraph:
            let children: [UInt32] = node.children.map { $0 }
            return .paragraph(children: children)

        case .blockquote:
            let children: [UInt32] = node.children.map { $0 }
            return .blockQuote(children: children)

        case .codeblock:
            guard let body = node.text else { return nil }
            return .codeBlock(info: node.codeInfo, body: body)

        case .list:
            let items: [[UInt32]] = node.listItems.map { item in
                item.children.map { $0 }
            }
            // ordered_start == -1 (default) means unordered (None).
            let orderedStart: UInt64? = node.orderedStart == -1 ? nil : UInt64(clamping: node.orderedStart)
            return .list(orderedStart: orderedStart, items: items)

        case .rule:
            return .rule

        case .emphasis:
            let children: [UInt32] = node.children.map { $0 }
            return .emphasis(children: children)

        case .strong:
            let children: [UInt32] = node.children.map { $0 }
            return .strong(children: children)

        case .inlinecode:
            guard let code = node.text else { return nil }
            return .inlineCode(code)

        case .link:
            let children: [UInt32] = node.children.map { $0 }
            return .link(children: children, href: node.href)

        case .image:
            guard let alt = node.alt else { return nil }
            // Image src is stored in the `url` field (Rust encoder: `args.url = src`).
            return .image(alt: alt, title: node.imgTitle, src: node.url)

        case .softbreak:
            return .softBreak

        case .hardbreak:
            return .hardBreak

        case .placeholder:
            let reason = decodePlaceholderReason(node.placeholderReason)
            return .placeholder(reason: reason)

        default:
            // Forward-compat: any unknown kind collapses to a depth_limit
            // placeholder rather than failing the whole decode (D1).
            return .placeholder(reason: .depthLimit)
        }
    }

    private static func decodeNostrUri(_ fbUri: nmp_content_WireNostrUri?) -> NostrWireUri? {
        guard let fbUri,
              let uri = fbUri.uri,
              let primaryId = fbUri.primaryId else { return nil }
        let kind: NostrWireUriKind
        switch fbUri.kind {
        case .profile:  kind = .profile
        case .event:    kind = .event
        case .address:  kind = .address
        default:        kind = .profile
        }
        let relays: [String] = fbUri.relays.compactMap { $0 }
        // event_kind sentinel: u32::MAX == UInt32.max means None.
        let eventKind: UInt32? = fbUri.eventKind == UInt32.max ? nil : fbUri.eventKind
        return NostrWireUri(
            uri: uri,
            kind: kind,
            primaryId: primaryId,
            relays: relays,
            author: fbUri.author,
            eventKind: eventKind
        )
    }

    /// MediaKind discriminant: Image=0, Video=1, Audio=2 (from Rust encoder).
    private static func decodeMediaKind(_ v: UInt8) -> NostrMediaKind {
        switch v {
        case 0: return .image
        case 1: return .video
        case 2: return .audio
        default: return .image
        }
    }

    /// InvoiceKind discriminant: Bolt11=0, Bolt12=1, Cashu=2 (from Rust encoder).
    private static func decodeInvoice(kind: UInt8, payload: String) -> NostrWireInvoice {
        switch kind {
        case 0: return .bolt11(payload)
        case 1: return .bolt12(payload)
        case 2: return .cashu(payload)
        default: return .bolt11(payload)
        }
    }

    private static func decodePlaceholderReason(_ reason: nmp_content_PlaceholderReason) -> NostrWirePlaceholderReason {
        switch reason {
        case .depthlimit:    return .depthLimit
        case .unresolveduri: return .unresolvedUri
        default:             return .depthLimit
        }
    }

    // ── Relation counts ───────────────────────────────────────────────────────

    private static func makeRelationCounts(_ fb: nmp_nip01_NoteRelationCounts) -> NoteRelationCounts? {
        guard let replies = fb.replies,
              let reactions = fb.reactions,
              let reposts = fb.reposts,
              let zaps = fb.zaps else { return nil }
        return NoteRelationCounts(
            replies: makeRelationCount(replies),
            reactions: makeRelationCount(reactions),
            reposts: makeRelationCount(reposts),
            zaps: makeRelationCount(zaps)
        )
    }

    private static func makeRelationCount(_ fb: nmp_nip01_RelationCount) -> RelationCount {
        switch fb.state {
        case .known:   return .known(fb.count)
        case .loading: return .loading
        default:       return .loading
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    /// Honour the schema's `has_*` companion bool: an absent field
    /// (`present == false`) is `nil` even when the string accessor returns "".
    private static func optionalString(_ value: String?, present: Bool) -> String? {
        present ? value : nil
    }
}
